File: app/routers/bid.py
from fastapi import APIRouter

from util.tables import Tables
from util.delete import delete
from util.update import update
from util.retrieve import retrieve
from util.insert import insert

from models import BidModel


router = APIRouter(prefix="/bids", tags=['bids'])

@router.get("/{bid_id}")
def get_bid(
    bid_id: int
):
    success, _, message, items = retrieve(
        table=Tables.Collector.value,
        single=True,
        bid_id=bid_id
    )

    return {"data": items[0], "success": success, "message": message}

@router.get("/")
def get_bids(
    bid_id: int | None = None,
    price: str | None = None,
    gt__price: str | None = None,
    lt__price: str | None = None, 
    auction_id: int | None = None,
    collector_id: int | None = None,
    payment_done: bool | None = None,
    created_at: str | None = None
):
    success, count, message, items = retrieve(
        table=Tables.Tag.value,
        single=False,
        bid_id=bid_id,
        price=price,
        gt__price=gt__price,
        lt__price=lt__price,
        auction_id=auction_id,
        collector_id=collector_id,
        payment_done=payment_done,
        created_at=created_at
    )

    return {"data": items, "success": success, "message": message, "count": count}


@router.post("/")
def create_new_bid(request_data: BidModel):
    success, message = insert(request_data)
    return {"message": message, "success": success}

@router.delete("/{bid_id}")
def delete_bid(bid_id: int):
    success, message = delete(
        table=Tables.Bid.value,
        bid_id=bid_id
    )
    return {"message": message, "success": success}


@router.put("/{bid_id}")
def update_bid(bid_id: int, request_data: BidModel):
    success, message = update(
        table=Tables.Bid.value,
        model=request_data,
        bid_id=bid_id
    )
    return {"message": message, "success": success}

File: frontend/api/bid.ts
import { get, post, put, deleteItem, toQueryString } from "./crude";

export const getBid = async (bid_id: number): Promise<ApiReuslt<Bid>> => {
  return get<Bid>(`http://localhost:8000/bids/${bid_id}`);
};

type BidQueryParams = {
  bid_id?: number;
  price?: string;
  gt__price?: string;
  lt__price?: string;
  auction_id?: number;
  collector_id?: number;
  payment_done?: boolean;
  created_at?: string;
};

export const getBids = async (
  params: BidQueryParams
): Promise<ApiReuslt<Bid[]>> => {
  const queryString = toQueryString(params);
  return get<Bid[]>(`http://localhost:8000/bids?${queryString}`);
};

type CreateBidData = {
  price: string;
  auction_id: number;
};

export const createNewBid = async (
  data: CreateBidData
): Promise<ApiReuslt<Bid>> => {
  return post<CreateBidData, Bid>(`http://localhost:8000/bids`, data);
};

export const deleteBid = async (bid_id: number): Promise<ApiReuslt<null>> => {
  return await deleteItem(`http://localhost:8000/bids/${bid_id}`, null);
};

export const acceptPayment = async (
  bid_id: number
): Promise<ApiReuslt<null>> => {
  return await post<{ bid_id: number }, null>(
    `http://localhost:8000/bids/accept_payment/${bid_id}`,
    { bid_id }
  );
};
